src/pricing/uniswap.ts

```diff
@@ -123,26 +126,74 @@ export function getTokenPrice(address: Address): BigDecimal {
       if (j == 0) {
         let native = getNativePrice();
         stable = stable.times(native);
-      } else if (j == 4) {
+      } else if (j == 5) {
         let eth = getEthPrice();
         stable = stable.times(eth);
       }
 
+      if (stable.lt(MIN_USD_PRICING)) {
+        continue;
+      }
+
       // compute price
-      if (stable.gt(MIN_USD_PRICING)) {
-        let token = ERC20.bind(address);
-        let amount = integerToDecimal(tokenReserve, BigInt.fromI32(token.decimals()));
+      let token = ERC20.bind(address);
+      let amount = integerToDecimal(tokenReserve, BigInt.fromI32(token.decimals()));
+
+      return stable.div(amount);
+
+    }
+  }
+
+  // try uniswap v3 pricing
+  if (block < UNISWAP_FACTORY_V3_START_BLOCK) {
+    return ZERO_BIG_DECIMAL;
+  }
+  let factory = UniswapFactoryV3.bind(Address.fromString(UNISWAP_FACTORY_V3));
+  let fees: number[] = [3000, 10000]; // 500];
+
+  // try each stable and fee tier
+  for (let i = 0; i < stables.length; i++) {
+    for (let j = 0; j < fees.length; j++) {
+      let poolAddress = factory.getPool(address, Address.fromString(stables[i]), fees[j] as i32);
 
-        return stable.div(amount);
+      if (poolAddress == zero) {
+        continue;
+      }
+
+      let stablePrice = BigDecimal.fromString('1.0');
+      if (i == 0) {
+        stablePrice = getNativePrice();
+      } else if (j == 5) {
+        stablePrice = getEthPrice();
+      }
+
+      let stableERC20 = ERC20.bind(Address.fromString(stables[i]));
+      let stableAmount = integerToDecimal(stableERC20.balanceOf(poolAddress), BigInt.fromI32(decimals[j] as i32));
+      if ((stableAmount.times(stablePrice)).lt(MIN_USD_PRICING)) {
+        continue;
       }
+
+      // compute price
+      let pool = UniswapPoolV3.bind(poolAddress);
+      let slot0 = pool.slot0();
+      let sqrtPriceX96 = slot0.value0;
+      let price = (sqrtPriceX96.times(sqrtPriceX96).toBigDecimal()).div(
+        BigInt.fromI32(2).pow(96 * 2).toBigDecimal() // effective bit shift >> (96*2)
+      );
```

Comment:
Are there docs or something I could look at that explain how on earth this calculation leads to a price?

Reply:
haha there are
https://docs.uniswap.org/sdk/guides/fetching-prices#understanding-sqrtprice